Decrease use of downcast<>() in rendering code

cdumez opened the pull request:
#### 8faf2e8d8d0c8f86f9b4833917927ef4d3808545
<pre>
Decrease use of downcast&lt;&gt;() in rendering code
<a href="https://bugs.webkit.org/show_bug.cgi?id=266831">https://bugs.webkit.org/show_bug.cgi?id=266831</a>

Reviewed by Brent Fulgham.

* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::inlineLogicalWidth):
(WebCore::shouldSkipWhitespaceAfterStartObject):
(WebCore::BreakingContext::handleReplaced):
(WebCore::BreakingContext::handleText):
(WebCore::BreakingContext::canBreakAtThisPosition):
(WebCore::BreakingContext::commitAndUpdateLineBreakIfNeeded):
* Source/WebCore/rendering/line/LineBreaker.cpp:
(WebCore::LineBreaker::skipLeadingWhitespace):
* Source/WebCore/rendering/line/LineInlineHeaders.h:
(WebCore::requiresLineBox):
(WebCore::setStaticPositions):
* Source/WebCore/rendering/mathml/MathMLStyle.cpp:
(WebCore::MathMLStyle::getMathMLStyle):
(WebCore::MathMLStyle::resolveMathMLStyleTree):
(WebCore::MathMLStyle::updateStyleIfNeeded):
(WebCore::MathMLStyle::resolveMathMLStyle):
* Source/WebCore/rendering/mathml/RenderMathMLFenced.cpp:
(WebCore::RenderMathMLFenced::updateFromElement):
* Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
(WebCore::RenderMathMLFraction::unembellishedOperator const):
* Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
(WebCore::toVerticalStretchyOperator):
* Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
(WebCore::RenderMathMLScripts::unembellishedOperator const):
(WebCore::RenderMathMLScripts::italicCorrection):
* Source/WebCore/rendering/mathml/RenderMathMLUnderOver.cpp:
(WebCore::horizontalStretchyOperator):
(WebCore::RenderMathMLUnderOver::hasAccent const):
(WebCore::RenderMathMLUnderOver::verticalParameters const):
* Source/WebCore/rendering/shapes/Shape.cpp:
(WebCore::Shape::createShape):
* Source/WebCore/rendering/shapes/ShapeOutsideInfo.cpp:
(WebCore::makeShapeForShapeOutside):

Canonical link: <a href="https://commits.webkit.org/272473@main">https://commits.webkit.org/272473@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5aa683cf2e5a9e1dd8011dfea27bfc79ee995727

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31743 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34237 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32531 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7674 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28337 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8795 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28336 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7586 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7755 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35584 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28698 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33871 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31727 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9500 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8520 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4148 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8362 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/mathml/MathMLStyle.cpp:
```diff
@@ -82,11 +82,10 @@ void MathMLStyle::updateStyleIfNeeded(RenderObject* renderer, MathMLElement::Mat
-            downcast<RenderMathMLToken>(renderer)->updateTokenContent();
+        auto* mathMLToken = dynamicDowncast<RenderMathMLToken>(renderer);
+        if (mathMLToken && !mathMLToken->isAnonymous())
+            mathMLToken->updateTokenContent();
```
Way easier to read now!

brentfulgham commented on Source/WebCore/rendering/shapes/Shape.cpp:
```diff
@@ -102,7 +102,7 @@ Ref<const Shape> Shape::createShape(const BasicShape& basicShape, const LayoutPo
 
     case BasicShape::Type::Circle: {
-        const auto& circle = downcast<BasicShapeCircle>(basicShape);
+        const auto& circle = uncheckedDowncast<BasicShapeCircle>(basicShape);
```
Out of curiosity: do these unchecked casts give any perf improvement?

cdumez replied:
They avoid a redundant type check. I am still trying to make the release type check affordable on Intel so I can enable it there.

brentfulgham reviewed (approved):
Looks good. Curious if this is perfect neutral, or if you see any benefit (aside from code clarity).

cdumez committed:
Decrease use of downcast<>() in rendering code
https://bugs.webkit.org/show_bug.cgi?id=266831

Reviewed by Brent Fulgham.

* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::inlineLogicalWidth):
(WebCore::shouldSkipWhitespaceAfterStartObject):
(WebCore::BreakingContext::handleReplaced):
(WebCore::BreakingContext::handleText):
(WebCore::BreakingContext::canBreakAtThisPosition):
(WebCore::BreakingContext::commitAndUpdateLineBreakIfNeeded):
* Source/WebCore/rendering/line/LineBreaker.cpp:
(WebCore::LineBreaker::skipLeadingWhitespace):
* Source/WebCore/rendering/line/LineInlineHeaders.h:
(WebCore::requiresLineBox):
(WebCore::setStaticPositions):
* Source/WebCore/rendering/mathml/MathMLStyle.cpp:
(WebCore::MathMLStyle::getMathMLStyle):
(WebCore::MathMLStyle::resolveMathMLStyleTree):
(WebCore::MathMLStyle::updateStyleIfNeeded):
(WebCore::MathMLStyle::resolveMathMLStyle):
* Source/WebCore/rendering/mathml/RenderMathMLFenced.cpp:
(WebCore::RenderMathMLFenced::updateFromElement):
* Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
(WebCore::RenderMathMLFraction::unembellishedOperator const):
* Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
(WebCore::toVerticalStretchyOperator):
* Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
(WebCore::RenderMathMLScripts::unembellishedOperator const):
(WebCore::RenderMathMLScripts::italicCorrection):
* Source/WebCore/rendering/mathml/RenderMathMLUnderOver.cpp:
(WebCore::horizontalStretchyOperator):
(WebCore::RenderMathMLUnderOver::hasAccent const):
(WebCore::RenderMathMLUnderOver::verticalParameters const):
* Source/WebCore/rendering/shapes/Shape.cpp:
(WebCore::Shape::createShape):
* Source/WebCore/rendering/shapes/ShapeOutsideInfo.cpp:
(WebCore::makeShapeForShapeOutside):

Canonical link: https://commits.webkit.org/272473@main
